src/datasets/load.py

```diff
@@ -407,11 +473,137 @@ def infer_module_for_data_files_in_archives(
+        config_data_dir = metadata_config.pop("data_dir", None)
+        config_base_path = os.path.join(base_path, config_data_dir) if config_data_dir else base_path
+        config_data_files_dict, config_patterns = get_data_files_locally(
+            base_path=config_base_path, data_files=config_data_files
+        )
+        metadata_config["data_files"] = config_data_files_dict
+
+        if config_data_files is None and supports_metadata and config_patterns != DEFAULT_PATTERNS_ALL:
+            update_data_files_with_metadata_files_locally(metadata_config["data_files"], base_path=config_base_path)
+
+
+def get_data_files_in_dataset_repository(
+    hfh_dataset_info, base_path: str, data_files: Optional[Union[str, List, Dict]] = None
+):
+    patterns = (
+        sanitize_patterns(data_files)
+        if data_files is not None
+        else get_data_patterns_in_dataset_repository(hfh_dataset_info, base_path)
+    )
+    data_files = DataFilesDict.from_hf_repo(
+        patterns, dataset_info=hfh_dataset_info, base_path=base_path, allowed_extensions=ALL_ALLOWED_EXTENSIONS
+    )
+    return data_files, patterns
+
+
+def update_data_files_with_metadata_files_in_dataset_repository(
+    hfh_dataset_info, data_files: DataFilesDict, base_path: str
+):
+    try:
+        metadata_patterns = get_metadata_patterns_in_dataset_repository(hfh_dataset_info, base_path=base_path)
+    except FileNotFoundError:
+        metadata_patterns = None
+    if metadata_patterns is not None:
+        metadata_files = DataFilesList.from_hf_repo(
+            metadata_patterns, dataset_info=hfh_dataset_info, base_path=base_path
+        )
+        for key in data_files:
+            data_files[key] = DataFilesList(
+                data_files[key] + metadata_files,
+                data_files[key].origin_metadata + metadata_files.origin_metadata,
+            )
+
+
+def get_data_files_for_metadata_configs_in_dataset_repository(
+    hfh_dataset_info, metadata_configs_dict: MetadataConfigs, base_path: str, supports_metadata: bool
+):
+    for metadata_config in metadata_configs_dict.values():
+        config_data_files = metadata_config.pop("data_files", None)
+        config_data_dir = metadata_config.pop("data_dir", None)
+        config_base_path = os.path.join(base_path, config_data_dir) if config_data_dir else base_path
+        config_data_files_dict, config_patterns = get_data_files_in_dataset_repository(
+            hfh_dataset_info, base_path=config_base_path, data_files=config_data_files
+        )
+        metadata_config["data_files"] = config_data_files_dict
+
+        if config_data_files is None and supports_metadata and config_patterns != DEFAULT_PATTERNS_ALL:
+            update_data_files_with_metadata_files_in_dataset_repository(
+                hfh_dataset_info, data_files=metadata_config["data_files"], base_path=config_base_path
+            )
+
+
+def get_builder_info_from_dataset_metadata(
```

Comment:
I'm not sure about the name `builder_info` because this function returns `DatasetInfo`.